addons/l10n_sa_edi/models/res_company.py

```diff
@@ -0,0 +1,202 @@
+import re
+import json
+from base64 import b64encode
+from odoo import models, fields, api, service
+from cryptography import x509
+from cryptography.x509 import ObjectIdentifier
+from cryptography.x509.oid import NameOID
+from cryptography.hazmat.backends import default_backend
+from cryptography.hazmat.primitives import hashes
+from cryptography.hazmat.primitives import serialization
+from cryptography.hazmat.primitives.serialization import Encoding, load_pem_private_key
+from cryptography.hazmat.primitives.asymmetric import ec
+
+CERT_TEMPLATE_NAME = b'\x0c\x12ZATCA-Code-Signing'
+
+
+class ResCompany(models.Model):
+    _inherit = "res.company"
+
+    l10n_sa_organization_unit = fields.Char("Organization Unit", copy=False,
+                                            help="The branch name for taxpayers. In case of VAT Groups, this field "
+                                                 "should contain the 10-digit TIN number of the individual group "
+                                                 "member whose device is being onboarded")
+
+    l10n_sa_serial_number = fields.Char("Serial Number", copy=False,
+                                        help="The serial number of the Taxpayer solution unit. Provided by ZATCA")
+
+    l10n_sa_private_key = fields.Binary(attachment=False, groups="base.group_erp_manager", copy=False,
+                                        compute="_l10n_sa_compute_private_key", store=True,
+                                        help="The private key used to generate the CSR and obtain certificates")
+    l10n_sa_csr = fields.Binary(attachment=False, copy=False, compute="_l10n_sa_compute_csr", store=True,
+                                help="The Certificate Signing Request that is submitted to the Compliance API")
+
+    l10n_sa_otp = fields.Char("OTP", copy=False, help="OTP required to get a CCSID. Can only be acquired through "
+                                                      "the Fatoora portal.")
+
+    l10n_sa_compliance_csid_json = fields.Char("CCSID JSON", copy=False,
```

Comment:
As per E-Invoicing Detailed Technical Guidelines, section 3.5.1
``` 
There should be one CSID per Taxpayer and
also one CSID per unique sequence of generated documents.
```
So does CSID fields belong to company or journal?

Reply:
Checked yesterday with ZATCA over a QnA session, the CSIDs have to be set on the journals in this case, and I have adapted the code accordingly.